lib/fourOhFour.js

```diff
@@ -0,0 +1,190 @@
+'use strict'
+
+const FindMyWay = require('find-my-way')
+
+const Reply = require('./reply')
+const Request = require('./request')
+const Context = require('./context')
+const {
+  kRoutePrefix,
+  kCanSetNotFoundHandler,
+  kFourOhFourLevelInstance,
+  kReply,
+  kRequest,
+  kContentTypeParser,
+  kBodyLimit,
+  kLogLevel,
+  kFourOhFourContext,
+  kFourOhFour,
+  kMiddlewares,
+  kHooks
+} = require('./symbols.js')
+const { beforeHandlerWarning } = require('./warnings')
+
+/**
+ * Each fastify instance have a:
+ * kFourOhFourLevelInstance: point to a fastify instance that has the 404 handler setted
+ * kCanSetNotFoundHandler: bool to track if the 404 handler has alredy been set
+ * kFourOhFour: the singleton instance of this 404 module
+ * kFourOhFourContext: the context in the reply object where the handler will be executed
+ */
+function fourOhFour (options) {
+  const { logger, modifyCoreObjects, genReqId } = options
+
+  // 404 router, used for handling encapsulated 404 handlers
+  const router = FindMyWay({ defaultRoute: fourOhFourFallBack })
+
+  const fof = {
+    router,
+    setNotFoundHandler: setNotFoundHandler,
+    setContext: setContext,
+    arrange404: arrange404
+  }
+
+  return fof
+
+  function arrange404 (instance) {
+    // Change the pointer of the fastify instance to itself, so register + prefix can add new 404 handler
+    instance[kFourOhFourLevelInstance] = instance
+    instance[kCanSetNotFoundHandler] = true
+  }
+
+  function basic404 (req, reply) {
+    reply.code(404).send(new Error('Not Found'))
+  }
+
+  function setContext (context) {
```

Comment:
It would be clearer if the Fastify instance was passed as an argument instead of being bound to `this`.

```suggestion
  function setContext (instance, context) {
```